doc/adr/0005-wego-core-auth-strategy.md

```diff
@@ -0,0 +1,83 @@
+  - Note that the documentation explicitly calls out that PKCE is optimal for client-side apps without a public cloud server (our use case)
+  - The documentation example specifies a `CLIENT_SECRET`, but that may be a documentation bug. A `CLIENT_SECRET` should not be necessary
+- [Bitbucket Implicit Grant](https://developer.atlassian.com/cloud/bitbucket/oauth-2/#2--implicit-grant--4-2-)
+  - No `CLIENT_SECRET` required here
+
+### Kubernetes Cluster
+
+All operations will need to go through an API server that is installed on the cluster via the `wego gitops install` command. This API server will run with a service account that provides the ability to do the following:
+
+- Create `Namespace` resources
+- Create `Application`, `Source`, and `Kustomize|Helm` resources
+- Read all Kubernetes objects, in all namespaces
+
+### Security Concerns
+
+Each of the Git Providers listed above support some sort of "personal access token" that allow for "impersonation" style authorization. These tokens never expire, and provide "sudo" access to the user's Github account.
+
+For this reason, we want to utilize the more short-lived OAuth tokens to avoid exposing a personal access token. We also don't want to persist these tokens longer than we have to.
+
+In addition, flows that require a `CLIENT_SECRET` will require us to embed it in the binary, which leaves us open to exposing it via decompilation.
+
+### UX Concerns
+
+Given that the Weave GitOps UI and API server may be running on a user's cluster, we may not always know the OAuth callback URL ahead of time. For this reason, we will need to use OAuth flows that do not require a static callback URLs. The flows listed above in the Git Providers section fit that constraint.
+
+We will need to ensure that the callback URL is configurable by the user via environment variables. This will allow them to run and expose the UI/API servers on whichever endpoint they choose.
+
+## Design
+
+Since most modern Git Providers will support a form of browser-based OAuth, **we will utilize the browser for both the CLI and UI authentication with Git Providers**.
+
+In the case of the CLI, we can utilize a short-lived browser session that will recieve the OAuth callback and pass the token to the CLI:
+
+![CLI Auth Diagram](cli_auth.svg)
+
+UI auth will work in a similar way, with a more straight-forward set of steps:
+
+![UI Auth Diagram](ui_auth.svg)
+
+## Decision
+
+Implement the flow(s) defined in the "Design" section
+
+## Consequences
+
+A SaaS application would be able to hide their `CLIENT_SECRET` behind their firewalled infrastructure. We do not have that luxury, so we will have to ship the `CLIENT_SECRET` with the binary. The only flow that requires this is the GitHub OAuth Flow, which gives us the alternative Device Flow specially built for our exact use case. We may be able to utilize their Device Flow as a one-off (since other providers do not have that feature).
```

Comment:
For acquiring a token, we don't need the client_secret.  Implicit grant, device flow, and PKCE don't require it.

Reply:
It says that right in the paragraph? 

> The only flow that requires this is the GitHub OAuth Flow, which gives us the alternative Device Flow specially built for our exact use case